query.py

```diff
@@ -0,0 +1,394 @@
+                    ]
+                ):
+                    attribute_url = (
+                        parent_rsc_attribute
+                        + ":"
+                        + child_rsc_type
+                        + "."
+                    )
+                    include_url = (
+                        parent_rsc_type
+                        + ":"
+                        + parent_rsc_attribute
+                        + ":"
+                        + child_rsc_type
+                    )
+
+                    self.rscs_graph.add_edge(
+                        parent_rsc_internal,
+                        child_dict[parent_rsc_attribute],
+                        attribute_parent=attribute_url,
+                        include=include_url,
+                    )
+                if (
+                    check
+                    in self.possible_references[child_rsc_type][
+                        "searchRevInclude"
+                    ]
+                ):
+                    attribute_url = (
+                        parent_rsc_type
+                        + ":"
+                        + parent_rsc_attribute
+                        + ":"
+                    )
+                    revinclud_url = (
+                        parent_rsc_type
+                        + ":"
+                        + parent_rsc_attribute
+                        + ":"
+                        + child_rsc_type
+                    )
+
+                    self.rscs_graph.add_edge(
+                        child_dict[parent_rsc_attribute],
+                        parent_rsc_internal,
+                        attribute_child=attribute_url,
+                        revinclude=revinclud_url,
+                    )
+                else:
+                    possibilities = self.possible_references[
+                        parent_rsc_type
+                    ]["searchInclude"]
+                    print(f"{check} not in {possibilities}")
+
+    def _where(self, **wheres):
+        """Builds the url_params that best respects the cumulative 
+        conditions specified on resource attributes
+        """
+        # With the current approach, adjustments must be made to remove 
+        # results that do not match the conditions (due to the fact that
+        # parameters chained to _has parameters are not cumulative).
+        max_conditions = -1
+        for internal_name in wheres.keys():
+            # the main resource chosen is the one with the most
+            # specified parameters. Choice to be discussed, here I think
+            # that it allows to make the best possible selection
+            # knowing that in the chained parameters and with _has are
+            # not cumulative.
+            curr_num_conditions = len(wheres[internal_name].keys())
+            if curr_num_conditions > max_conditions:
+                max_conditions = curr_num_conditions
+                self.main_internal_name = internal_name
```

Comment:
I don't get it. The fhir resource on which the query is done is the one having the most "conditions"?
The naming still doesn't help here, I think (condition? internal_name?)

Reply:
One thing that needs to be determined in order to build the query url is which resource will be selected to replace the `type` in the search query (`GET [base]/[type]?name=value&...{&_format=[mime-type]}`)

For the moment (and I'm sure this choice can be improved) I choose the resource that has the most where conditions specified in the configuration file.

I made this choice because the first level of parameters is cumulative:

For example:
```
GET Encounter?type=OKI&status=in-progress&reason-code=1372004
```
will return Encounters of type OKI and with status "in progress" and a reason-code of 1372004 ([url to check it on api](http://hapi.fhir.org/baseR4/Encounter?type=OKI&status=in-progress&reason-code=1372004))

On the other hand, `Chained parameters` are not cumulative:
``
GET Patient?general-practitioner.name=Joe&general-practitioner.address-state=MN
``

> may return Patients cared for by Joe from CA and Jane from MN: no one practitioner need satisfy both conditions. E.g. the chains are evaluated separately. ([FHIR](https://www.hl7.org/fhir/search.html#chaining))

it is also the same with `Reverse Chaining`.

Reply:
Maybe `num_param_conditions` would be clearer than `curr_num_conditions`? or `num_where_conditions`?

Reply:
Don't you think it would be better if the user had to specify explicitly the main resource on which he wants to perform the query in our query language?
I'm not even sure this custom rule will always be true